#29019 Migrate from Karma to Jest  `libs/ui` and `libs/utils`

oidacra opened the pull request:
### Checklist
- [x] Tests
- [ ] Translations
- [ ] Security Implications Contemplated (add notes if applicable)
![CleanShot 2025-02-24 at 13 11 08@2x](https://github.com/user-attachments/assets/46e40af0-0187-4587-a234-2badced68743)



### Additional Info
** any additional useful context or info **
This pull request introduces several significant changes across multiple files to improve commit message standards, update unit testing rules, and migrate from Karma to Jest for testing. The most important changes are summarized below:

### Commit Message Standards

* [`.cursor/rules.json`](diffhunk://#diff-19f43daae59c1d52302173155e4605f696866f65a0e34013eac4e80261a1eff2R1-R74): Introduced new rules for commit message patterns, scopes, and types to enforce a consistent format for commit messages.

### Unit Testing Rules

* [`.cursor/rules/dotcms-unit-test-rules.mdc`](diffhunk://#diff-9589a57120d8702699741cd58a94eb78948890cfeca5a3887e377dfd0fd2f457R1-R274): Added comprehensive unit testing rules, including best practices for file structure, element selection, and user-centric testing guidelines for DotCMS Angular projects.

### Testing Framework Migration

* [`core-web/libs/ui/jest.config.ts`](diffhunk://#diff-f378d35375c3870fac2d2b41219a885a6644a409c1f1aecc50533a781a4ae742R1-R24): Added Jest configuration for the `ui` library, including setup files, transformation rules, and coverage directory settings.
* [`core-web/libs/ui/karma.conf.js`](diffhunk://#diff-4135d40b37d522e0b9d10105c924545c6bf93318ec8bac5438ac3603d2580757L1-L16): Removed Karma configuration as part of the migration to Jest.
* [`core-web/libs/ui/project.json`](diffhunk://#diff-9808513312178f14b6faff1e47fa984207931404c99727d55de75f8575b0afa8L13-R22): Updated the project configuration to use Jest for running tests instead of Karma.

### Test File Updates

* [`core-web/libs/ui/src/lib/components/add-to-bundle/dot-add-to-bundle.component.spec.ts`](diffhunk://#diff-30c7579837230bdeac077664d70be9379f1c0794d03ae3cab9ba014f7ca66e53L53-R54): Replaced `spyOn` with `jest.spyOn` for mocking functions in tests. [[1]](diffhunk://#diff-30c7579837230bdeac077664d70be9379f1c0794d03ae3cab9ba014f7ca66e53L53-R54) [[2]](diffhunk://#diff-30c7579837230bdeac077664d70be9379f1c0794d03ae3cab9ba014f7ca66e53L82-R82) [[3]](diffhunk://#diff-30c7579837230bdeac077664d70be9379f1c0794d03ae3cab9ba014f7ca66e53L153-R153) [[4]](diffhunk://#diff-30c7579837230bdeac077664d70be9379f1c0794d03ae3cab9ba014f7ca66e53L168-R168)
* [`core-web/libs/ui/src/lib/components/dot-action-menu-button/dot-action-menu-button.component.spec.ts`](diffhunk://#diff-f1995df579d6adc97a57512a0ca8ccfd080816904c977e240efe8633ac44d219L1-R1): Updated import statements to use `@ngneat/spectator/jest` and replaced `spyOn` with `jest.spyOn`. [[1]](diffhunk://#diff-f1995df579d6adc97a57512a0ca8ccfd080816904c977e240efe8633ac44d219L1-R1) [[2]](diffhunk://#diff-f1995df579d6adc97a57512a0ca8ccfd080816904c977e240efe8633ac44d219L143-R149) [[3]](diffhunk://#diff-f1995df579d6adc97a57512a0ca8ccfd080816904c977e240efe8633ac44d219L241-R261)
* [`core-web/libs/ui/src/lib/components/dot-ai-image-prompt/ai-image-prompt.component.spec.ts`](diffhunk://#diff-c6fe7f58329266e68ab51a7ad0026f7884e26dd7097758feacc403fc5b524490L7-R33): Updated import statements and replaced `spyOn` with `jest.spyOn` for mocking functions in tests. [[1]](diffhunk://#diff-c6fe7f58329266e68ab51a7ad0026f7884e26dd7097758feacc403fc5b524490L7-R33) [[2]](diffhunk://#diff-c6fe7f58329266e68ab51a7ad0026f7884e26dd7097758feacc403fc5b524490L39-R49) [[3]](diffhunk://#diff-c6fe7f58329266e68ab51a7ad0026f7884e26dd7097758feacc403fc5b524490R58-R67) [[4]](diffhunk://#diff-c6fe7f58329266e68ab51a7ad0026f7884e26dd7097758feacc403fc5b524490L66-R81)

This PR fixes: #29019

oidacra committed:
refactor(utils): migrate test framework from Karma to Jest

- remove test.ts file used by Karma
- update tsconfig.spec.json to use Jest types
- add proper getComputedStyle mock in test-setup.ts
- update test configuration in project.json

oidacra committed:
refactor(ui): migrate UI library tests from Karma to Jest

- Update test configuration in project.json to use Jest
- Replace Spectator imports from '@ngneat/spectator' to '@ngneat/spectator/jest'
- Replace Jasmine spies with Jest mocks
- Update test setup to use Jest-compatible mocking techniques
- Remove Karma-specific configuration files

oidacra committed:
feat(ui): Add comprehensive unit testing rules and guidelines

- Created a new rules file for DotCMS Angular unit testing
- Defined best practices for testing with Jest, Spectator, and @testing-library/angular
- Established guidelines for file structure, element selection, and user-centric testing
- Introduced data-testid conventions and testing utilities
- Provided template examples for component, form, and dialog testing

oidacra committed:
test(ui): Enhance DotKeyValueTableRowComponent unit tests

- Improve test coverage for key-value table row component
- Add focus event handling for input fields
- Update test cases for hidden field and password visibility
- Remove unnecessary imports and refactor test assertions
- Ensure comprehensive testing of component interactions

oidacra committed:
chore(ui): Add ESLint disable comments for test files

Disable specific ESLint rules in test files to improve test writing flexibility:
- Suppress @typescript-eslint/no-explicit-any
- Suppress @typescript-eslint/no-unused-vars
- Suppress @typescript-eslint/no-empty-function
- Suppress @typescript-eslint/ban-ts-comment

This change allows more lenient type checking and unused variable handling in test files, making test writing and mocking easier.

oidacra committed:
chore(ui): fix format

nicobytes commented on .cursor/rules/dotcms-unit-test-rules.mdc:
```diff
@@ -0,0 +1,274 @@
+const button = fixture.debugElement.query(By.css('.submit-button'));
+
+// ✅ Use data-testid
+const button = screen.getByTestId('submit-button');
```
@oidacra Actually, in unit tests, we use `spectator.query(byTestId('id'))` or just `byTestId('id')`, but from spectator

oidacra replied:
ohh good catch. Let me add that.

oidacra replied:
Improved

nicobytes commented on .cursor/rules/dotcms-unit-test-rules.mdc:
````diff
@@ -0,0 +1,274 @@
+});
+```
+
+## User-Centric Testing Guidelines
````
Amazing!

oidacra committed:
docs(testing): Update unit testing guidelines with best practices

- Correct spelling of Playwright
- Emphasize data-testid and element selection best practices
- Add guidance for CSS class verification
- Improve code examples with more explicit recommendations
- Enhance formatting and clarity of testing rules

oidacra committed:
Merge branch 'main' into 29019-from-karma-to-jest-ui

oidacra committed:
Merge branch 'main' into 29019-from-karma-to-jest-ui
